fixes and refactoring start

martinetd opened the pull request:
- first few commits are small fixes, probably OK. might want to push them as a separate PR to merge and stop worrying about it so the MainActivity.kt diff is clean when I start with it (soon)
- last two commits start refactoring: for now I've just added plenty of types for SipMessage, plenty of todos, and a few questions

I've got plenty to talk about here but it's getting late and for now the only thing that matters is what you think of these types so we can iterate; I'd rather not to build too much on a house of cards.

If types are ok, next is the big questions I've left in comments regarding how to parse, the mix of BufferedReader and BytesArray for message body doesn't work well... But we don't care about the bodies yet, so assuming types are ok I could start using this in MainActivity.kt, or start digging at a network layer.

martinetd committed:
launchIms: make callId constant

using a different callId on second register was ignored by kddi

martinetd committed:
launchIms: REGISTER response parsing: use first sip: url for subscribe

kddi provides two sip addresses and stock rom uses the first one as preferred
endpoint. Remember first one instead of overwriting.
Ultimately that might need to be configurable instead as it might depend on
carriers...

martinetd committed:
gradle: update to 7.4.2

newer gradle is required for java 17.

Update:
 - gradle distributionUrl in gradle-wrapper.properties
 - gradle and kotlin-gradle-plugin in build.gradle
 - remove jcenter repository that is getting deprecated and isn't
actually used for this project

martinetd committed:
launchIms: de-constify imei

martinetd committed:
client thread: repeat Call-ID in reply

martinetd committed:
add me.phh.sip with parse helpers and tests

martinetd committed:
SipMessage: add serialization

phhusson reviewed (commented):
What does those two URLs look like?

phhusson commented on app/src/main/java/me/phh/sip/SipMessage.kt:
```diff
@@ -0,0 +1,119 @@
+data class SipCommonMessage(
+	val firstLine: String,
+	val headers: Map<String, List<SipHeader>>,
+	// TODO: messageBody needs to handle multipart, including nested multipart,
```
Have you seen it used yet?

martinetd replied:
the RFC used it for INVITE and I blindly trusted it was used... but now I'm checking it's not for me, although I've only tried calling outside and have no idea what the server would send me when called from outside

martinetd replied:
FWIW here's what rfc5621 has to say about it:

[4.2](https://datatracker.ietf.org/doc/html/rfc5621#section-4.2).  Mandatory Support for 'multipart' Message Bodies
```
   For all MIME-based extensions to work, the recipient needs to be able
   to decode the multipart bodies.  Therefore, SIP UAs MUST support
   parsing 'multipart' MIME bodies, including nested body parts.
   Additionally, UAs MUST support the 'multipart/mixed' and 'multipart/
   alternative' MIME types.  Support for other MIME types such as
   'multipart/related' is OPTIONAL.

      Note that, by default, unknown 'multipart' subtypes are treated as
      'multipart/mixed'.  Also note that SIP extensions can also include
      'multipart' MIME bodies in responses.  That is why both UACs and
      UASs need to support 'multipart' bodies.

   Legacy SIP UAs without support for 'multipart' bodies generate a 415
   (Unsupported Media Type) response when they receive a 'multipart'
   body in a request.  A UAC sending a 'multipart' body can receive such
   an error response when communicating with a legacy SIP UA that
   predates this specification.

      It has been observed in the field that a number of legacy SIP UAs
      without support for 'multipart' bodies simply ignored those bodies
      when they were received.  These UAs did not return any error
      response.  Unsurprisingly, SIP UAs not being able to report this
      type of error have caused serious interoperability problems in the
      past.
```

I think the actual multipart parsing can be left for when we actually need it, but instead of having users of sip messages explicitly check for content-type (99% won't), just directly built it as a map with content-type as key so if they get something of unexpected type it won't work automatically?

e.g. `val body: Map<String, ByteArray>`

Pro:
 - reasonably easy to use
 - makes checking we're parsing what we expect mandatory

Con:
 - won't allow producing nested multipart if required by some extension
 - won't allow sending multiple parts with same content type

phhusson commented on app/src/main/java/me/phh/sip/SipMessage.kt:
```diff
@@ -0,0 +1,119 @@
+  		.takeWhile { it.isNotBlank() }
+		.fold(emptyMap<String, List<SipHeader>>(), fold@{ headers, line ->
+			// ignore anything we don't recognize
+			// TODO: if line starts with a space, content should be happened to previous value
```
that's sadly a requirement for me, I have services who send multi-line Authorization header...

martinetd replied:
Ok. That's a bit tricky with the current line by line code, it'd be easier if we do the split ourselves so e.g. we split on \r\n but not \r\n<space> for this function -- I think it's the cleanest thing to do to handle ByteArrays in the messages anyway so I'll do that next :/

phhusson commented on app/src/main/java/me/phh/sip/SipMessage.kt:
```diff
@@ -0,0 +1,119 @@
+ */
+private val splitHeader = "^(.+?) *: *(.+)$".toRegex()
+
+fun parseHeaders(input: BufferedReader): Map<String, List<SipHeader>> =
```
I presume that unlike me, you read the specs and you did it for a reason, but aren't usually headers just Map<String, SipHeader>? Some keys can be repeated multiple times in a header and have a meaning?

phhusson replied:
nvm, there is `P-Associated-Uri` already

martinetd replied:
Some headers like `Contact`, `Route` etc are allowed to be provided multiple times, and either:
```
Contact: first contact, second contact
```
or
```
Contact: first contact
Contact: second contact
```
have the same meaning (now I'm checking Concact can have parameters, so my comment about never mixing parameters and commas was wrong, but it's now clear to me the parameters belong to the comma-segment so there's no ambiguity)

... and I get multiple Route headers which are defined as a comma separated list, so we need to at least faithfully remember them all. As far as I can see it's just about repeating the same Route: headers in our messages, so the current code should be fine, but if we're not going to use it we could just build a big Route: firstone, secondone for our replies if that seems easier.
I've preferred to think the other way around and split it out if we get commas for headers we care about so users won't need to worry if it's a single one or multiple when they use it.

RFC3261 also specifies Authorization can be repeated, but I didn't see it in practice:
```
   Although not a comma-
   separated list, this header field name may be present multiple times,
   and MUST NOT be combined into a single header line using the usual
   rules described in [Section 7.3](https://datatracker.ietf.org/doc/html/rfc3261#section-7.3).
```

phhusson commented on app/src/main/java/me/phh/sip/SipMessage.kt:
```diff
@@ -0,0 +1,119 @@
+		?.getOrNull(0)
+		?.value?.toInt()
+		?.let(::ByteArray)
+	// XXX BROKEN ... and now we're here, there's no way to read into a ByteArray from a BufferedReader,
```
I know the feeling... been there... done that...

phhusson commented on app/src/main/java/me/phh/sip/SipMessage.kt:
```diff
@@ -117,3 +117,16 @@ fun parseMessage(input: BufferedReader): SipMessage? {
 	}
 }
+
+fun serializeMessage(message: SipCommonMessage): ByteArray =
```
I must be blind, but I can't see how you send first line -_-'

martinetd replied:
It's the original accumulator value to the later fold, line 130.
I had it in the first fold originally as that made more sense, but that lead to a leading \r\n so I had to move it last -- thinking of fold like string join bit me...

martinetd commented:
> What does those two URLs look like?

```
P-Associated-Uri: <sip:+818012341234@ims.mnc051.mcc440.3gppnetwork.org>  
P-Associated-Uri: <sip:+818012341234@ims.imsnw.kddi.ne.jp>  
```

It probably doesn't matter much, but my client then proceeds using the first one while your code originally overwrote it and used the later

martinetd commented:
> That's all fine by me, I can probably merge it as-is since it doesn't seem you broke anything to me

As for this (how to move forward): I'm fine either way.

It doesn't look like I'll need to do huge changes to the current code, so I think building incrementally instead of continuously rewriting history makes more sense -- I'm generally careful not to break things (e.g. introduce new interfaces then use them) so I agree it's probably fine to merge and will make further reviews easier so that's probably the way to go, but I can just keep piling up here as you prefer.

martinetd committed:
SipMessage: parseMessage: return common message if first line was not identified properly

This should not matter much in practice but will allow logging
to be handled by the application

martinetd commented:
One more question: I noticed I'm not using the same style (tab vs space etc) at all as in MainActivty.kt, and found some `ktfmt` tool.

It probably doesn't matter all that much at this point but should we get a run of that in before growing the code too much?
(I don't use an IDE but that can be enforced at commit time or github action if we agree to it)

Unfortunately it looks like your style wasn't what they expect either, but it's somewhat close enough with the `--kotlinlang-style` switch

phhusson commented:
I use android-studio so it's mostly using Android Studio's formatting
Ok for `ktfmt --kotlinglang-style`.

It looks like using ktfmt in Android Studio requires a bit of manual setting, and there doesn't seem to be some generic file header that text editors could read to know which ktfmt setting to use?

martinetd commented:
No idea, I'm running good ol' vim with little integration :/
As long as it's something I can run on git commit I don't really care what we use, or if we use nothing at all - I just like to have some automation to catch tabs when we're not using them, and since I'm not familiar with kotlin I wanted to have something confirm where standard line breaks happen etc (I tried using when () -> with a block and wondered what was appropriate for that when I looked for ktfmt)

android studio seems to support .editorconfig so perhaps just that'd be enough?

martinetd committed:
SipMessage: ktfmt pass

martinetd committed:
MainActivity: ktfmt pass

martinetd committed:
add editorconfig

This will ensure I don't keep adding tabs everywhere...

martinetd committed:
SipReader: add a reader implementation

We want sort of line-based parsing while still being able to read
ByteArrays, so cannot use BufferedReader, but BufferedInputStream
does not have any line oriented parsing.

Implement our own parsing, and as bonus have it handle line continuations

martinetd committed:
SipReader: add readNBytes method

we'll use readNBytes to make sure full message is read

martinetd committed:
SipMessage: use SipReader

martinetd commented:
I've:
 - added a SipReader class over BufferedInputStream (I didn't notice the difference between BufferedReader and BufferedInputStream that you were using in MainActivity.kt... I wasn't getting why you had to reimplement `lines()` but it all makes sense now :|)
 - That SipReader is a bit more specialized:
   - it trims carriage returns like bufferedReader's
   - it stops on empty line automatically, this isn't 100% needed but makes for something easier to use and we need empty line handling anyway for the next point:
   - it peeks at next lines to handle continuations at this level, this is safe for SIP as we know there's more to read
 - switched SipMessage to use it
 - added editorconfig, that'll be good enough for now. I did pass ktfmt once just to see and comitted that for MainActivity, but I'd be happy to remove that commit, I don't really care as long as tab/spaces aren't mixed.

So now my SipMessages should be good enough for you as well (line continuations got there), I'll switch MainActivity to use it next and then will look at separating network out.

martinetd committed:
Sip*Test: fix binary handling and add empty message tests

- convert test data to ByteArray: Having +String(body) would corrupt body
for bytes > 127, and we were converting these to ByteArrays anyway for
inputStream-ification

- add a couple of tests for totally empty message (-> null message),
and standalone crlf that have been seen for keepalives (also null message,
still allowing whatever is next to be a valid message)

martinetd committed:
SipMessage: split headers on comma and handle parameters

martinetd committed:
SipMessage: move away from data class to allow autofill

martinetd committed:
SipMessage: handle headers compact form

martinetd committed:
rename package from me.phh.sip to me.phh.ims

could not import me.phh.sip from MainActivity.kt -- need to look up the proper way

martinetd committed:
SipMessage: Do not split parameters out by default

always splitting parameters made handling heavy and is not actually required for
most headers.
Instead provide a getParams function that can be used when required; this will
also be usable for Authorization later

martinetd committed:
SipMessage: serialize allow/supported as single line

while spec allows headers to be one header per line, it's a waste of packet
space and it's not what official clients are sending

martinetd committed:
SipMessage: simplify Request/Response constructor

do not require a full first line for constructor, just use destination
or status string

martinetd committed:
SipMessage: add toString and bubble it up to request/reponse

martinetd committed:
SipMessages: rework types again: make base SipMessage bear common fields

having base SipMessage as abstract allows accessing firstLine headers or body without cast
and is much nicer to use

martinetd committed:
SipMessage parameters: add getAuthValues() like getParams()

share code in a private splitParams function.
Can probably do better, but that'll do for now

martinetd committed:
SipMethod: add missing NOTIFY method. Might be missing others.

martinetd committed:
SipMessage: share toByteArray in abstract class

martinetd committed:
SipMessage: take all words for SipResponse

for non-ok status the response might be multiple words, take a slice

martinetd committed:
fixup! SipMessage: share toByteArray in abstract class

martinetd committed:
Add SipChallenge helper and move some utils functions

martinetd committed:
MainActivity.kt: trivial refactoring using new types

This doesn't change any logic except not closing the client connection
after a single message

martinetd commented:
ok that's about it for the first pass using new types -- the logic is mostly untouched so I don't think I broke anything, but tcp still doesn't work well for me so I didn't get anything past the register itself.

next would be breaking things out a step further, but I'm still undecided on how to abstract the network away.. I guess I'll just try something and see how it goes...

If you have time for a new review, now'd be a good time to take a fresh look at the type changes as it's evolved quite a bit since last week, but I hope I've kept it in a general direction you're comfortable with? Please tell me if there's something you'd see differently.

martinetd committed:
MainActivity: reindent badly indented header strings

martinetd committed:
Security-Client: send aes-cbc and properly handle 'q' in reply

martinetd committed:
SipMessage: prefix message toString conversion with '> '

make sure we don't use toString for anything sent on the wire
by prefixing each line with an angle bracket.

Also remove the abstract class toString implementation that
wasn't used as it wasn't inherited to remove duplicate code

martinetd committed:
hmac_sha1: use 160 bits padded key

Link: https://android-review.googlesource.com/c/platform/frameworks/base/+/2002572
Link: https://www.arib.or.jp/english/html/overview/doc/STD-T63v9_10/5_Appendix/Rel5/33/33203-5c0.pdf

pihug12 commented:
Sorry if it's not the right place to ask, but is there still some works done on this? Or should I follow this somewhere else?

Many thanks for this effort :slightly_smiling_face:

martinetd commented:
hi @pihug12 -- I'm still working on it (a bit); the /e/ foundation is financing development so the "active" branches have moved to their gitlab which isn't public; and this branch has been more or less left as is since.

https://github.com/martinetd/android_ims has up-to-date-ish developments; but it now needs to be built in-tree because the IMS API isn't available to normal builds, and it's still nowhere to complete (the address for outgoing SMS isn't always correct and we have no idea where one could get it from, calls aren't implemented at all, the app isn't properly backgrounded...)
If you can or know someone who can help on any of these issues we'd be happy to delegate some of it and explain what's needed to get started :)
